flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBufferAccessor.java

```diff
@@ -235,31 +235,32 @@ public void lockNode(final NodeId node) {
 	 * @throws Exception Thrown if the system cannot access the state.
 	 */
 	public void releaseNode(final NodeId node) throws Exception {
-		Lockable<SharedBufferNode> sharedBufferNode = sharedBuffer.getEntry(node);
-		if (sharedBufferNode != null) {
-			if (sharedBufferNode.release()) {
-				removeNode(node, sharedBufferNode.getElement());
-			} else {
-				sharedBuffer.upsertEntry(node, sharedBufferNode);
+		// the stack used to detect all nodes that needs to be released.
+		Stack<NodeId> nodesToExamine = new Stack<>();
+		nodesToExamine.push(node);
+
+		while (!nodesToExamine.isEmpty()) {
+			NodeId curNode = nodesToExamine.pop();
+			Lockable<SharedBufferNode> curBufferNode = sharedBuffer.getEntry(curNode);
+
+			if (curBufferNode == null) {
+				break;
```

Comment:
What does it mean if the `currBufferNode` is null? When can it occur? Shouldn't we rather `continue` to clear all elements from `nodesToExamine`?

Reply:
@dawidwys Good question. Thanks.

There can be duplicated attempts to remove a node. So if we do not check whether currBufferNode is null, an NPE will be thrown. 

As an example, please take a look at CEPITCase#testFlatSelectSerializationWithAnonymousClass. There are two attempts to release the node from NFA#doProcess method. 

When a node is released (and set to null), the decedent nodes have already been released too, so there is no need to  continue examining nodes in nodesToExamine.

Please note that such logic is not designed by me. The original code worked the same way. I just refactor the code so that the logic is identical to the original code.